### PR TITLE
Implement the software Master interface and transaction routing

### PR DESCRIPTION
**Pull request by ljanyst**
This CL implements a software master and a software RAM slave test and all the necessary infrastructure to route transactions between the two.

### CODE REVIEW

File: src/common/Data.hh [hunk 1]
```diff
@@ -85,4 +91,24 @@ private:
...
+ */
+enum class TransactionType { READ_REQ, WRITE_REQ, READ_RESP, WRITE_RESP };
+
+/**
+ * Transaction
+ */
+struct Transaction {
+    TransactionType type = TransactionType::READ_REQ;  //!< Type of the transaction
+    uint64_t initiator = 0;  //!< The IP block that initiated the transaction
+    uint64_t target = 0;  //!< The IP block that projcessed the transaction; set by the router
```
Comment (jfrohnhofen) [inline src/common/Data.hh]: type: processed

File: src/router/sw_axi/router.go [hunk 2]
```diff
@@ -47,17 +91,122 @@ func NewRouter(uri string, numClients int) (*Router, error) {
...
+		}
+
+		if msg.Type() != wire.TypeTRANSACTION {
+			log.Fatalf("Received unexpected message: %s", msg.Type())
+		}
+
+		txn := msg.Txn(nil)
+		op := "Write"
+		if txn.Type() == wire.TransactionTypeREAD_RESP || txn.Type() == wire.TransactionTypeREAD_REQ {
+			op = "Read "
```
Comment (jfrohnhofen) [inline src/router/sw_axi/router.go:176]: This has a space at the end while "Write" has not. Not sure if this is significant.

**Review by jfrohnhofen — APPROVED** [review]

File: src/router/sw_axi/router.go (hunk 2 above)
Comment (ljanyst) [reply src/router/sw_axi/router.go:176]: Yeah, the printout looks nicer when read and write (words) have equal width :)

File: src/common/Data.hh (hunk 1 above)
Comment (ljanyst) [reply src/common/Data.hh]: Thanks!

**Comment by ljanyst** [comment]
@jfrohnhofen Thanks for having a very careful look :)

**Review by Richard-Gordon — APPROVED** [review]

**Pull request merged**